test/unit/test/tmp_filename_test.cpp

```diff
@@ -67,13 +67,25 @@ TEST(tmp_filename_dtr, dtr)
     EXPECT_FALSE(fs::exists(path.parent_path()));
 }
 
-// throw if invalid TMPDIR
+// throw if TMPDIR not writeable
 TEST(tmp_filename_throw, throw)
 {
-    if (!system("man putenv > /dev/null 2>&1"))
-    {
-        char str[] = "TMPDIR=/invalid";
-        putenv(str);
-        EXPECT_THROW(seqan3::test::tmp_filename t1{"throw"}, std::filesystem::filesystem_error);
-    }
+    // Create a directory in the temporary directory.
+    std::filesystem::path test_directory = std::filesystem::temp_directory_path();
+    test_directory /= "tmp_filename_throw";
+    std::filesystem::create_directory(test_directory);
+    // Remove write permissions.
+    std::filesystem::permissions(test_directory,
+                                 std::filesystem::perms::owner_write,
+                                 std::filesystem::perm_options::remove);
+    // Set TMPDIR. This is the first env var that is looked up for `temp_directory_path` inside the `tmp_filename`.
+    setenv("TMPDIR", test_directory.c_str(), 1); // name, value, overwrite
+
+    EXPECT_THROW(seqan3::test::tmp_filename t1{"throw"}, std::filesystem::filesystem_error);
+
+    // Add write permissions (needed for deletion) and delete the test directory.
+    std::filesystem::permissions(test_directory,
+                                 std::filesystem::perms::owner_write,
+                                 std::filesystem::perm_options::add);
+    std::filesystem::remove(test_directory);
```

Comment:
Ok, this is not a trivial code block by far.
Here are some issues I detected: 

The return type of create_directory is not evaluated (a clear library issue by not declaring the code as [[nodiscard]]). Hence, what ever is tested next is not valid anymore.
Second, setting the permissions can throw. In this case if the first one throws, the resource (directory) is not removed. That can happen btw anytime inside of the code block.
Also the `remove` return type is not evaluated.

I suggest to put the creation of the directory and the handling of the file permissions in an explicit RAII wrapper class for this test only. It should have an explicit release function that returns the state of remove and tests that the result is true.
The RAII wrapper class should attempt to free the resource on destruction using the no-throw version.
The entire block is put in a try-catch block and the catch block always FAILS() with a dedicated error message and the hint that there is probably a leaking resource in the directory with the offending path name.

Reply:
Maybe it would also be good to ensure, that the directory did not exist before. Otherwise, other processes might get interrupted when this test removes the write permissions.

Reply:
So basically the RAII wrapper class throws as well on construction and the error is handled by the catch block making it a failing test not a terminated test.